python/ray/serve/_private/deployment_state.py

```diff
@@ -1389,6 +1390,268 @@ def __repr__(self):
         return repr(self._replicas)
 
 
+class DeploymentRankManager:
+    """Manages replica ranks for a deployment.
+    This class handles rank assignment, release, consistency checking, and reassignment.
+    It maintains the rank system invariants and provides a clean interface for rank operations.
+    """
+
+    def __init__(self, _fail_on_error: Optional[bool] = None):
+        # Maps replica_id to assigned rank
+        self._replica_ranks: Dict[str, int] = {}
+        # Set of available ranks (initially empty, grows as target replicas change)
+        self._available_ranks: Set[int] = set()
+        # Next rank to assign (increments as new replicas are created)
+        self._next_rank: int = 0
+        # Whether to fail on rank errors (for testing control)
+        self._fail_on_error = (
+            _fail_on_error
+            if _fail_on_error is not None
+            else RAY_SERVE_FAIL_ON_RANK_ERROR
+        )
+
+    def assign_rank(self, replica_id: str) -> int:
+        """Assign a rank to a new replica.
+        Args:
+            replica_id: The unique ID of the replica
+        Returns:
+            The assigned rank
+        Raises:
+            RuntimeError: If the replica already has a rank assigned
+        """
+        if replica_id in self._replica_ranks:
+            raise RuntimeError(
+                f"Replica {replica_id} already has a rank assigned: {self._replica_ranks[replica_id]}"
+            )
+
+        # First try to reuse an available rank
+        if self._available_ranks:
+            rank = min(self._available_ranks)
+            self._available_ranks.remove(rank)
+        else:
+            # Otherwise use the next available rank
+            rank = self._next_rank
+            self._next_rank += 1
+
+        self._replica_ranks[replica_id] = rank
+        return rank
+
+    def release_rank(self, replica_id: str) -> None:
+        """Release a rank when a replica is stopped.
+        Args:
+            replica_id: The unique ID of the replica whose rank should be released
+        """
+        if replica_id not in self._replica_ranks:
+            raise RuntimeError(f"Replica {replica_id} has no rank assigned")
+
+        rank = self._replica_ranks.pop(replica_id)
+        self._available_ranks.add(rank)
+
+    def recover_rank(self, replica_id: str, rank: int) -> None:
```

Comment:
this seems similar to the port recovery issue doyoung is working on. will this also be called as part of the control loop?